### DISCUSSION TITLE
Do you use the filename guesswork feature?

### DISCUSSION THREAD

**Discussion by jonaswinkler** [jonaswinkler/paperless-ng#83] [General]
You can name files in certain ways and paperless will pick up metadata from that, such as tags, dates and titles.

See guesswork in the documentation.

Do you use it?

**Comment by brianwanderson** [comment]
Hi,

Here is some feedback on how I use paperless:

I really like the  PAPERLESS_FILENAME_DATE_ORDER=YMD setting in the config file.  I prefix all of my documents with YYYY.MM.DD so that I end up with file names like: "2020.12.01 - Document Title.pdf".  That way, I can always sort my documents chronologically in a file browser.  It's really convenient that paperless will automatically set the documents date from the file name on import.

Beyond grabbing the date, I don't use any of the filename guesswork features.  Personally, I think anything more is overkill.  I don't want to be bothered by having to remember some special file name format like "Correspondent - Title - tag,tag,tag.pdf"    It's easier for me to name my files however I wish and to assign correspondents and tags in the GUI after import.

**Comment by qcasey** [comment]
Basically the same here, I'd see myself using filename guesswork on dates when bulk importing a bunch of documents. Paperless has been very good at filling in titles and dates otherwise. 

I personally don't use correspondents or tags at all. The search function is really good enough.

**Comment by rknightion** [comment]
Personally no, as my scanner gives random filenames and going through and renaming would be more effort when the existing search and auto tagging capabilities are so good 👍

**Comment by jonaswinkler** [comment]
I'll probably reduce that to date detection only at some point.

**Reply by DocLambda** [reply]
Hey Jonas,

I started using (and coding for) paperless-ng some days ago and didn't know about the filename-guesswork feature. Is there a reason to remove it? If it's not a lot of code or hard to maintain, I'd rather keep/evolve existing features as there are always people that find exactly this useful.

Thanks for the good work! Keep it up! Hope to submit my first PR soon...

Best regards,

    Sven

**Reply by jonaswinkler** [reply]
> I started using (and coding for) paperless-ng some days ago and didn't know about the filename-guesswork feature. Is there a reason to remove it? If it's not a lot of code or hard to maintain, I'd rather keep/evolve existing features as there are always people that find exactly this useful.

* I see that it was useful for a couple things, but most of the time, it got in the way of users that did not know about this feature. Lots of documents where some part of the original file name was used as the correspondent (wrong), and cut off titles. Lots of manual effort to fix that.
* The implementation was VERY strict about the patterns. It was also pretty much impossible to add any kind of user configurable patterns, since everything was hard coded.
* Then there's also the PAPERLESS_DATEORDER (or something) option, that also parses dates from filenames.
* If, at all, I'd like to make metadata parsing from filenames user configurable and the implementation should also take this DATEORDER thing into account.

> Thanks for the good work! Keep it up! Hope to submit my first PR soon...

What are you working on?

**Reply by DocLambda** [reply]
How about allowing the user to

1. Use/don't use the feature by enable/disable flag?
2. Allow the user to specify his/her own pattern for parsing the filename (maybe even conditionally)?

> If, at all, I'd like to make metadata parsing from filenames user configurable and the implementation should also take this DATEORDER thing into account.

So seems like we can agree on this. :-) Awesome!

> What are you working on?

I do have a PR (almost) ready for showing a swagger UI for the paperless-ng API. This might be helpful for developers that want to play with the API to e.g. write a small program to upload a document from their scanner (guess who might want to do that :-)).
Furthermore, I've implemented a user-definable meta-data model but I'm currently struggling with the frontend integration. My idea is to use something like [this](https://github.com/jsoma/kull) to directly derive tags/metadata from the document. I do have a lot of invoices that I want to archive and like to extract the date, amount, etc. from the document and store it as (searchable) metadata.
Ideas are on using django-haystack for search (to be able to scale), cloud storage (hey we might want to bring back file-encryption for this), reducing the time-to-awesome (I had a hard time to set this up, my target is to do `pip install paperless-ng`, `paperless-ng start`) and user-definable processing pipelines. :-)
I love coding (if time permits) so in case there is some urgent work to be done... :-)

**Reply by jonaswinkler** [reply]
> I do have a PR (almost) ready for showing a swagger UI for the paperless-ng API. This might be helpful for developers that want to play with the API to e.g. write a small program to upload a document from their scanner (guess who might want to do that :-)).

Interesting. The Django REST framework also provides a user-browsable API, with filtering and all that, so I'm curious how that compares. Sounds redundant right now, but lets see.

> Furthermore, I've implemented a user-definable meta-data model but I'm currently struggling with the frontend integration.

Curious, this is on my todo list and I've already got pretty definite ideas about how to do it. See https://github.com/jonaswinkler/paperless-ng/issues/274#issuecomment-766062277, and the rest of that thread as well. Is your current progress available somewhere?

> Ideas are on using django-haystack for search (to be able to scale),

I'm using whoosh directly, without haystack, since certain search features (more like this and faceting) are not available through haystack.

> cloud storage (hey we might want to bring back file-encryption for this),

Propose a good plan about how to do it that also takes the database, search index and filenames into account and does not store the decryption key on the server. I've concluded that it's pretty much impossible without redesigning major parts of the application.

> my target is to do `pip install paperless-ng`, `paperless-ng start`

Paperless depends on external non-python binaries. The docker setup has a pretty good *time-to-awesome*, download some files, pull, go.

**Reply by DocLambda** [reply]
I'm probably hijacking this discussion but well... :-)

> The Django REST framework also provides a user-browsable API.

That's what I used to provide the OpenAPI specification directly derived from the django url-router. On top of it I added an Angular component to integrate swagger-ui.
![integrate swagger-ui](https://user-images.githubusercontent.com/11217448/106123310-11e36c00-615a-11eb-8f50-8cd7159ac0d8.png)

I have to cleanup the Angular-mess I made and try to submit this as a PR next week (hopefully).

> #274 

I will comment there.

> haystack

Let's discuss this later. Maybe I can try to make those features available through haystack... :-)

> Propose a good plan about how to do it that also takes the database, search index and filenames into account and does not store the decryption key on the server. I've concluded that it's pretty much impossible without redesigning major parts of the application.

Well to be honest this is not necessary in my case. What I want is to put the document **files** in the cloud storage but keep the database, index, etc. locally. This way I only need to "protect" the files but can keep the rest plain-text.

> Paperless depends on external non-python binaries. The docker setup has a pretty good time-to-awesome, download some files, pull, go.

Well true, but for one I don't like docker too much as a "solution to setup things that are too complicated otherwise" and secondly I hate to remember which commands to run when developing and starting the server, the consumer, the worker queue... Don't get me wrong, paperless-ng is awesome, but I had a hard time to start this from a `git clone` even when reading your documentation. Especially the starting of a (development) server instance is too hard. But I'm convinced wrapping this is not too complicated...

Thanks for all your work!

**Reply by jonaswinkler** [reply]
> Well to be honest this is not necessary in my case. What I want is to put the document **files** in the cloud storage but keep the database, index, etc. locally. This way I only need to "protect" the files but can keep the rest plain-text.
> 

Let's discuss this in #40 .

> Don't get me wrong, paperless-ng is awesome, but I had a hard time to start this from a `git clone` even when reading your documentation.

I know that getting the development environment up and running isn't exactly straight forward and that the documentation is severly lacking in that regard.

**Reply by DocLambda** [reply]
Cool. Will try to work on it a bit and let you know.

**Reply by jonaswinkler** [reply]
> That's what I used to provide the OpenAPI specification directly derived from the django url-router. On top of it I added an Angular component to integrate swagger-ui.

Regarding this, I don't think the average user needs to access the API directly, so integrating that into the front end is questionable.

**Comment by RosaEinhorn** [comment]
Is there a feature to add tags based on the name of a file in the /consume folder? I cannot find it. :-(

**Reply by jonaswinkler** [reply]
There is not. The previous implementation of that was not pretty good and was causing many issues with miss-named tags, correspondents, and incomplete titles.

**Comment by tanderson1992** [comment]
I know the guesswork feature is already gone, but I'll add that I used it extensively. I found it easy to bulk rename documents (years of bank statements, for example) with the proper names to be consumed into the correct correspondents, titles, dates, and tags. As I mentioned in another discussion, I thought my wife would balk, but after a few tries she was regularly scanning documents with her phone and naming them the correct way to be consumed. We had an occasional hiccup, but for the most part it worked well for us.

**Comment by dennisdek** [comment]
Hi Jonas,
this would be exactly what I'm looking for at the moment, but it seems it has been removed based on the replies above?
My use case: My scanner allows up to 10 FTP profiles. So my ideal would be to setup several profiles, all leading to the same consume folder - only difference is the filename that is generated for the scan.
E.g. a profile 'tax', another one 'school'. That way I can just select the right profile in the scanner and the document gets tagged correctly based on the file name, without any further action in the app itself.
